File: LeetCode/cpp/2815_Max_Pair_Sum_in_an_Array.cpp
class Solution {
public:
	int maxSum(vector<int>& nums) {
		int answer = -1;

		unordered_map<int, vector<int>> um;

		for (int n : nums) {
			string s = to_string(n);

			int maxDigit = 0;
			for (int i = 0; i < s.length(); i++) {
				if (maxDigit < (s[i] - '0')) {
					maxDigit = s[i] - '0';
				}
			}

			um[maxDigit].push_back(n);
		}

		for (auto m : um) {
			if (m.second.size() == 1) {
				continue;
			}

			vector<int> v = m.second;

			sort(v.begin(), v.end());

			answer = max(answer, v[v.size() - 1] + v[v.size() - 2]);
		}

		return answer;
	}
};

File: LeetCode/java/2815_Max_Pair_Sum_in_an_Array.java
import java.util.*;

class Solution {
    public int maxSum(int[] nums) {
        int answer = -1;

        Map<Integer, List<Integer>> map = new HashMap<>();

        for (int n : nums) {
            String s = String.valueOf(n);

            int maxDigit = 0;
            for (int i = 0; i < s.length(); i++) {
                if (maxDigit < (s.charAt(i) - '0')) {
                    maxDigit = s.charAt(i) - '0';
                }
            }

            if (!map.containsKey(maxDigit)) {
                map.put(maxDigit, new ArrayList<>());
            }

            map.get(maxDigit).add(n);
        }

        for (int key : map.keySet()) {
            if (map.get(key).size() == 1) {
                continue;
            }

            List<Integer> list = map.get(key);

            Collections.sort(list);

            answer = Math.max(answer, list.get(list.size() - 1) + list.get(list.size() - 2));
        }

        return answer;
    }
}